csrc/scheduler/mma_utils.cpp

```diff
@@ -703,11 +704,51 @@ std::unordered_set<IterDomain*> getMmaDomainSet(
   return {mma_domains.begin(), mma_domains.end()};
 }
 
+bool isConsumerAllocationInnerIDProducerAllocationInnerID(
+    TensorView* consumer) {
+  // We are only looking at the allocation domains accross
+  // the producer and consumer of a LoadStoreOp
+  auto LdStOp = consumer->definition();
+  NVF_CHECK(dynamic_cast<LoadStoreOp*>(LdStOp));
+
+  // Get all the IDs from the innermost ID of the allocation domain of
+  // the consumer to the root domain of the consumer.
+  auto vals = DependencyCheck::getAllValsBetween(
+      {consumer->getRootDomain().begin(), consumer->getRootDomain().end()},
+      {consumer->getMaybeAllocationDomain().back()});
+
+  auto ids = ir_utils::filterByType<IterDomain>(vals);
+  auto idsOnPath = std::vector<IterDomain*>(ids.begin(), ids.end());
+  NVF_CHECK(!idsOnPath.empty());
+  // This gives us the ID in the consumer root domain.
+  // We'll later map this ID to one in the producer.
+  auto idInConsumerRoot = idsOnPath.front();
+
+  auto producerOfLoadStore = LdStOp->input(0);
+  NVF_CHECK(producerOfLoadStore->isA<TensorView>());
+  PairwiseRootDomainMap mapAccrossLdStOp(
+      static_cast<TensorView*>(producerOfLoadStore), consumer);
+
+  auto c2pMap = mapAccrossLdStOp.mapConsumerToProducer();
+  auto prodID = c2pMap.at(idInConsumerRoot);
+
+  // If the innermost ID of the (maybe)Allocation domain
+  // is not the same as the mapped ID in the producer, then
+  // we need to transpose.
+  return static_cast<TensorView*>(producerOfLoadStore)
+             ->getMaybeAllocationDomain()
+             .back() == prodID;
+}
+
 } // namespace
 
 void WarpMmaSwizzler::scheduleLdMatrix(TensorView* tv, MmaOperand operand) {
-  bool transpose = tv->definition()->as<LoadStoreOp>()->opType() ==
-      LoadStoreOpType::LdMatrixTranspose;
+  bool transpose = !isConsumerAllocationInnerIDProducerAllocationInnerID(tv);
+  // We don't have to set the Op type here if we removed LdMatrix and LdMatrix
+  // and handle the decision to transpose or not in codegen.
+  tv->definition()->as<LoadStoreOp>()->setOpType(
+      transpose ? LoadStoreOpType::LdMatrixTranspose
+                : LoadStoreOpType::LdMatrix);
```

Comment:
Should we just remove `LdMatrixTranspose` and only keep `LdMatrix`, and in lowering, we use `isConsumerAllocationInnerIDProducerAllocationInnerID` to decide whether `LdMatrix` should be lowered as a `ldmatrix` or `ldmatrix.trans`? I think this task itself can be peeled out from this PR and merged independently as a separate PR.

Reply:
I'll try and send a stacked PR on top of this.

Reply:
I think it should be on the bottom of this PR, where without this PR, the producer "AllocationInnerID" will be just the producer rfactor inner id.

Reply:
Basically I am looking for splitting this PR as smaller PRs, and I believe this part is a good choice to get out.